File: src/csharp/Foundgine.Core/Semantic/Query/SemanticQueryOptions.cs
namespace Foundgine.Core.Semantic.Query;

/// <summary>
/// Protocol-neutral query controls, including filters, ordering, limits,
/// offsets, and forward cursor pagination.
/// </summary>
public sealed record SemanticQueryOptions(
    SemanticFilterExpression? Filter = null,
    IReadOnlyList<SemanticOrderTerm>? Order = null,
    int? Limit = null,
    int? Offset = null,
    string? After = null)
{
    public IReadOnlyList<SemanticOrderTerm> EffectiveOrder => Order ?? [];

    public bool HasCursor => !string.IsNullOrWhiteSpace(After);
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/query/SemanticQueryOptions.java
package com.foundgine.core.semantic.query;

import java.util.*;

public record SemanticQueryOptions(SemanticFilterExpression filter, List<SemanticOrderTerm> order, Integer limit,
		Integer offset, String after) {
	public SemanticQueryOptions {
		order = order == null ? List.of() : List.copyOf(order);
	}

	public SemanticQueryOptions() {
		this(null, List.of(), null, null, null);
	}

	public List<SemanticOrderTerm> effectiveOrder() {
		return order;
	}

	public boolean hasCursor() {
		return after != null && !after.isBlank();
	}
}
